src/python/gem5/components/devices/gpus/amdgpu.py

```diff
@@ -0,0 +1,279 @@
+from .viper_shader import ViperShader
+
+
+class BaseViperGPU(SubSystem):
+    _base_pci_dev = 8
+    _gpu_count = 0
+    _my_id = 0
+
+    @classmethod
+    def next_pci_dev(cls):
+        cls._gpu_count += 1  # Use count for this particular type
+        return cls._base_pci_dev + cls._gpu_count - 1
+
+    @classmethod
+    def get_gpu_count(cls):
+        return cls._gpu_count
+
+    def __init__(self):
+        # Setup various PCI related parameters
+        self._my_id = self.get_gpu_count()
+        pci_dev = self.next_pci_dev()
+
+        device = AMDGPUDevice(pci_func=0, pci_dev=pci_dev, pci_bus=0)
+        self._device = device
+
+    def set_shader(self, shader: ViperShader):
+        self._shader = shader
+
+    def get_cpu_dma_ports(self):
+        return self._shader.get_cpu_dma_ports()
+
+    def connectGPU(self, board: AbstractBoard) -> None:
+        # Connect a CPU pointer. This is only used for SE mode. Any CPU will
+        # work, so pick assuming there is at least one
+        cpus = board.get_processor()
+        self._shader.set_cpu_pointer(cpus.cores[0].core)
+
+        # Connect all PIO buses
+        self._shader.connect_iobus(board.get_io_bus())
+
+        # The System() object in gem5 has a memories parameter which defaults
+        # to Self.all. This will collect *all* AbstractMemories and connect to
+        # the CPU side. To avoid this we manually assign the memories param to
+        # the CPU side memories. We need the MemInterface which is called dram
+        # in the MemCtrl class even though it might not be modelling dram.
+        memory = board.get_memory()
+        cpu_abs_mems = [mem.dram for mem in memory.get_memory_controllers()]
+        board.memories = cpu_abs_mems
+
+        # Make the cache hierarchy. This will create an independent RubySystem
+        # class containing only the GPU caches with no network connection to
+        # the CPU cache hierarchy.
+        self._device.gpu_caches = ViperGPUCacheHierarchy(
+            tcp_size=self._tcp_size,
+            tcp_assoc=self._tcp_assoc,
+            sqc_size=self._sqc_size,
+            sqc_assoc=self._sqc_assoc,
+            scalar_size=self._scalar_size,
+            scalar_assoc=self._scalar_assoc,
+            tcc_size=self._tcc_size,
+            tcc_assoc=self._tcc_assoc,
+            tcc_count=self._tcc_count,
+            cu_per_sqc=self._cu_per_sqc,
+            num_memory_channels=self._num_memory_channels,
+            cache_line_size=self._cache_line_size,
+            shader=self._shader,
+        )
+
+        # Collect GPU memory controllers created in the GPU cache hierarchy.
+        # First assign them as a child to the device so the SimObject unproxy.
+        # The device requires the memories parameter to be set as the system
+        # pointer required by the AbstractMemory class is set by AMDGPUDevice.
+        self._device.mem_ctrls = self._device.gpu_caches.get_mem_ctrls()
```

Comment:
The memory controllers should not be part of the cache. They should be a `MemorySystem` just like on the CPU side.

Reply:
The memory system is part of the device though. I'd have to look at what MemorySystem is and see if that causes problem with what one of the goals here is (to have the gpu be a standalone device you can create in one single line with some defaults assumed)